### ISSUE TITLE
custom script after cloudsql database creation

### ISSUE THREAD

**Issue by nrobert13** [GoogleCloudPlatform/k8s-config-connector#51]
would be nice to have a way to run custom scripts once the sql database is created, for:
- DB schema initialization

**Comment by xiaobaitusi** [comment]
Hi Robert, we don't yet support those sorts of imperative one-offs commands/scripts; our priority right now is expanding our RESTful resource coverage. 
Our recommendation for now would be to use the regular API. Keep it open and see if this is a common case for other customers and we will re-prioritize give the feedback.

**Comment by kibbles-n-bytes** [comment]
(Just brainstorming). One thought I had was us offering an image that that could be used as an [init container](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/), where you would specify the Config Connector resources the pod would want to be ready, and it'd watch those and block until everything's up. And then after initialization, the pod would run whatever custom image that does the work you'd want to happen. The only change would be that that pod's Kubernetes service account would need RBAC permission to read the Config Connector resources. 

How does that idea sound? We wouldn't be able to get to something like that until post-Q4, though.

**Comment by kibbles-n-bytes** [comment]
@nrobert13 What would you imagine to be a proper solution? I figure that this is a case not unique to SQL databases, and so it wouldn't be as useful to build something specific to this case if we could build something generic that addresses the overall need.

The overall issue I see is: in a declarative Kubernetes-native GitOps pipeline, how do you ensure that certain work happens after certain cloud resources are available? It may be one-off work that needs to happen (i.e. schema initialization), or it may be a persistent application with cloud dependencies, like our bookstore and musicians samples.

In both these cases, I'd imagine the desired work would be inside of a Kubernetes Pod. And since the standard way of blocking pods from doing their main work is through initialization containers, that would be the mechanism we'd want to hook into.

Happy to consider alternatives, though; didn't mean for it to sound like we were just pitching a workaround.

**Comment by nrobert13** [comment]
@kibbles-n-bytes, I was thinking about an additional field in the spec of the `sqldatabase` where you can provide a container to run and do the job once the database is ready.Basically this is kinda same thing as the initcontainer but tied to the database creation, in case multiple services access the database, you'd need to add the initcontainer to each of them ....

**Comment by kibbles-n-bytes** [comment]
Ah, I see. I'd say in this case it could be more general, and perhaps be an annotation on any resource specifying the image that a bespoke controller picks up and runs a job for with basic defaulted settings. We try to keep spec reserved for fields that correspond to input fields on the underlying resource.

In the case of multiple services accessing the database that don't gracefully handle the database not being up yet, yeah, you'd need to add the init container to each of them. Though I think that's the case even in your proposal. The init container in this case isn't doing the initialization of the database, it'd just be blocking until the database exists.

There's a separate question of "how do you block the services until the job that does necessary initialization work runs", though. I think in that case some explicit orchestration logic on top (e.g. one of: 2-staged deployment pipeline, Deployment Manager, Terraform, k8s operator, or a bespoke k8s controller) may make sense.
